crates/yakui-wgpu/src/samplers.rs

```diff
@@ -1,30 +1,47 @@
 pub(crate) struct Samplers {
-    nearest: wgpu::Sampler,
-    linear: wgpu::Sampler,
-    nearest_linear: wgpu::Sampler,
-    linear_nearest: wgpu::Sampler,
+    nearest_nearest_nearest: wgpu::Sampler,
+    linear_linear_nearest: wgpu::Sampler,
+    nearest_linear_nearest: wgpu::Sampler,
+    linear_nearest_nearest: wgpu::Sampler,
+    nearest_nearest_linear: wgpu::Sampler,
+    linear_linear_linear: wgpu::Sampler,
+    nearest_linear_linear: wgpu::Sampler,
+    linear_nearest_linear: wgpu::Sampler,
 }
 
 impl Samplers {
     pub fn new(device: &wgpu::Device) -> Self {
         use wgpu::FilterMode::{Linear, Nearest};
 
         Self {
-            nearest: sampler(device, Nearest, Nearest),
-            linear: sampler(device, Linear, Linear),
-            nearest_linear: sampler(device, Nearest, Linear),
-            linear_nearest: sampler(device, Linear, Nearest),
+            nearest_nearest_nearest: sampler(device, Nearest, Nearest, Nearest),
+            linear_linear_nearest: sampler(device, Linear, Linear, Nearest),
+            nearest_linear_nearest: sampler(device, Nearest, Linear, Nearest),
+            linear_nearest_nearest: sampler(device, Linear, Nearest, Nearest),
+            nearest_nearest_linear: sampler(device, Nearest, Nearest, Linear),
+            linear_linear_linear: sampler(device, Linear, Linear, Linear),
+            nearest_linear_linear: sampler(device, Nearest, Linear, Linear),
+            linear_nearest_linear: sampler(device, Linear, Nearest, Linear),
         }
     }
 
-    pub fn get(&self, min: wgpu::FilterMode, max: wgpu::FilterMode) -> &wgpu::Sampler {
+    pub fn get(
+        &self,
+        min: wgpu::FilterMode,
+        max: wgpu::FilterMode,
+        mipmap: wgpu::FilterMode,
+    ) -> &wgpu::Sampler {
         use wgpu::FilterMode::{Linear, Nearest};
 
-        match (min, max) {
-            (Nearest, Nearest) => &self.nearest,
-            (Linear, Linear) => &self.linear,
-            (Nearest, Linear) => &self.nearest_linear,
-            (Linear, Nearest) => &self.linear_nearest,
+        match (min, max, mipmap) {
+            (Nearest, Nearest, Nearest) => &self.nearest_nearest_nearest,
+            (Linear, Linear, Nearest) => &self.linear_linear_nearest,
+            (Nearest, Linear, Nearest) => &self.nearest_linear_nearest,
+            (Linear, Nearest, Nearest) => &self.linear_nearest_nearest,
+            (Nearest, Nearest, Linear) => &self.nearest_nearest_linear,
+            (Linear, Linear, Linear) => &self.linear_linear_linear,
+            (Nearest, Linear, Linear) => &self.nearest_linear_linear,
+            (Linear, Nearest, Linear) => &self.linear_nearest_linear,
```

Comment:
This code is very straightforward (and I'm so happy that Rust verifies this for us!) but... man, the combinatorics here are rough. 😅